src/core/Object3D.js

```diff
@@ -53,20 +106,68 @@ class Object3D extends EventDispatcher {
 		const quaternion = new Quaternion();
 		const scale = new Vector3( 1, 1, 1 );
 
-		function onRotationChange() {
+		const matrix = new Matrix4();
+		const matrixWorld = new Matrix4();
+
+		const self = this;
+		let rotationNeedsUpdate = false;
+		let quaternionNeedsUpdate = false;
 
-			quaternion.setFromEuler( rotation, false );
+		function onWrite() {
+
+			self.matrixNeedsUpdate = true;
 
 		}
 
-		function onQuaternionChange() {
+		function onRotationWrite() {
 
-			rotation.setFromQuaternion( quaternion, undefined, false );
+			self.matrixNeedsUpdate = true;
+			rotationNeedsUpdate = true;
 
 		}
 
-		rotation._onChange( onRotationChange );
-		quaternion._onChange( onQuaternionChange );
+		function onRotationRead() {
+
+			if ( quaternionNeedsUpdate === true ) {
+
+				quaternionNeedsUpdate = false;
+				rotation.setFromQuaternion( quaternion, undefined );
+				rotationNeedsUpdate = false;
+
+			}
+
+		}
+
+		function onQuaternionWrite() {
+
+			self.matrixNeedsUpdate = true;
+			quaternionNeedsUpdate = true;
+
+		}
+
+		function onQuaternionRead() {
+
+			if ( rotationNeedsUpdate === true ) {
+
+				rotationNeedsUpdate = false;
+				quaternion.setFromEuler( rotation );
+				quaternionNeedsUpdate = false;
+
+			}
+
+		}
+
+		function onMatrixChange() {
+
+			self.matrixWorldNeedsUpdate = true;
+
+		}
+
+		listenToProperties( position, [ 'x', 'y', 'z' ], undefined, onWrite );
+		listenToProperties( rotation, [ 'x', 'y', 'z', 'order' ], onRotationRead, onRotationWrite );
+		listenToProperties( quaternion, [ 'x', 'y', 'z', 'w' ], onQuaternionRead, onQuaternionWrite );
+		listenToProperties( scale, [ 'x', 'y', 'z' ], undefined, onWrite );
+		listenToMethods( matrix, onMatrixChange );
```

Comment:
`listenToProperties` adds an on-read and on-write callback to named keys (not a deopt), which we use here to flip `matrixNeedsUpdate` when transform properties are modified. On-read is used to convert quaternion/rotation like before, but only if their counterpart is dirty when read. If users stick to `quaternion`, they will avoid this conversion entirely. This is nice so we don't have to modify math classes like before.

`listenToMethods` I have as a special case for the local matrix whose members would be deopt if I just listened to all elements in its array. This is because its accessors are not named and unstable, so v8 would need to create a layer of indirection to make them stable. I instead reduce this to flipping `matrixWorldNeedsUpdate` whenever the local matrix is modified by its methods, and don't handle the case where `matrix.elements` is mutated directly. This keeps us on the fast path for common usage without breaking APIs who expected a 1-1 relationship between local and world matrix updates.